src/utils/bitfield.rs

```diff
@@ -0,0 +1,231 @@
+use core::ops::{Bound, Range, RangeBounds};
+
+/// Trait for manipulating subsets of integers.
+pub trait BitField {
+    const TYPE_SIZE: usize;
+
+    /// Return a boolean representing the state of the idx'th bit.
+    ///
+    /// # Arguments
+    ///
+    /// * `idx` - The index of the bit, 0 being the rightmost one.
+    ///
+    /// # Panics
+    ///
+    /// This method will panics if the given index is out of range.
+    fn get_bit(self, idx: usize) -> bool;
+
+    /// Return a value representing the state of a range of bits.
+    ///
+    /// # Arguments
+    ///
+    /// * `range` - The range of bits, starting from 0 for the rightmost one.
+    ///
+    /// # Panics
+    ///
+    /// This method will panics if the given range is invalid.
+    fn get_bits<T: RangeBounds<usize>>(self, range: T) -> Self;
+
+    /// Set the idx'th bit.
+    ///
+    /// # Arguments
+    ///
+    /// * `idx` - The index of the bit, 0 being the rightmost one.
+    /// * `value` - A boolean representing the desired state of the bit.
+    ///
+    /// # Panics
+    ///
+    /// This method will panics if the given index is out of range.
+    fn set_bit(self, idx: usize, value: bool) -> Self;
+
+    /// Set a range of bits to the given value.
+    ///
+    /// # Arguments
+    ///
+    /// * `range` - The range of bits, starting from 0 for the rightmost one.
+    /// * `value` - An integer representing the desired state of the bits.
+    ///
+    /// # Panics
+    ///
+    /// This method will panics if the given range is invalid.
+    /// Value must also not be greater in terms of bits than the given range.
+    fn set_bits<T: RangeBounds<usize>>(self, range: T, value: Self) -> Self;
+}
+
+macro_rules! impl_bitfields {
```

Comment:
I'm not entirely sure about the necessity of this. I see why this has been done but I can't decide wether this is a good or bad thing.

Reply:
I have to say I knew this comment was coming.
The thing is I do not see the point of duplicating code.
Tell me what is your decision on this and I can adapt.

Reply:
I don't see any easy solution, the best one from a syntax point of view would probably be a proc macro but there's two problems:
- We cannot derive native traits (at least I remember remote derivation needs a wrapper struct)
- The overall implementation of a proc macro is tedious

Let's keep it this way for now until we find a good solution, we can update it later in another PR.